Migrated content-editable directive from AngularJS to Angular

aditya993388 opened the pull request:
This PR migrates the content-editable directive from CoffeeScript/AngularJS to TypeScript/Angular.

Changes:
- Created a new Angular directive implementation using TypeScript
- Implemented ControlValueAccessor interface to maintain ngModel binding
- Added the directive to Angular module declarations
- Downgraded it for use in AngularJS components
- Removed the old CoffeeScript implementation

aditya993388 committed:
migrated content-editable directive to Angular, removed old coffeescript

martindolores commented on src/app/doubtfire-angular.module.ts:
Need to include in declarations

martindolores commented on src/app/doubtfire-angularjs.module.ts:
```diff
@@ -495,3 +495,7 @@ DoubtfireAngularJSModule.directive(
+DoubtfireAngularJSModule.directive(
+  'contenteditable',
+  downgradeComponent({ component: ContentEditableDirective })
+);
```
This wont be needed - as this is a directive, this cannot be downgraded

aditya993388 replied:
Hi @martindolores Thank you so much for the feedback.

- I have removed the downgradeComponent usage, as suggested. It was very helpful to understand that since this is a directive and not a component, it doesn’t need to be downgraded for AngularJS.

- ContentEditableDirective is now properly included in the declarations array of the Angular module.

-  I also updated the HTML templates where it was previously used by replacing ng-model with [(ngModel)] and explicitly adding contenteditable="true" to ensure it works correctly with Angular’s forms API.

martindolores reviewed (changes requested):
Looking good so far! Just a couple things to note down. 

As this is a **directive** and not a component, this cannot be downgraded to Angular JS. This will have to be just be used for Angular 17+ components

aditya993388 committed:
UPDATE: Finalise ContentEditable directive migration after review feedback

aditya993388 commented:
Hi @martindolores Thank you so much for the feedback.

 - I have removed the downgradeComponent usage, as suggested. It was very helpful to understand that since this is a directive and not a component, it doesn’t need to be downgraded for AngularJS.

- ContentEditableDirective is now properly included in the declarations array of the Angular module.

- I also updated the HTML templates where it was previously used by replacing ng-model with [(ngModel)] and explicitly adding contenteditable="true" to ensure it works correctly with Angular’s forms API.

martindolores commented on src/app/tasks/task-ilo-alignment/modals/task-ilo-alignment-modal/task-ilo-alignment-modal.tpl.html:
This wont work as the new content editable is not downgraded, will have to leave the .coffee file in for now until this component is migrated to angular 17+

martindolores commented on src/app/units/modals/unit-ilo-edit-modal/unit-ilo-edit-modal.tpl.html:
Same here. need to wait until this is migrated to angular 17+

martindolores commented on src/app/common/content-editable/content-editable.coffee:
Have to leave this one in as the new content-editable you have created cannot be downgraded

martindolores reviewed (changes requested):
Just a couple more things :)

ibi420 commented:
Hello @aditya993388, In this task, you have migrated the old contenteditable directive from AngularJS to Angular by creating a new directive file (content-editable.directive.ts) using ControlValueAccessor for [(ngModel)] support. I updated two template files (unit-ilo-edit-modal.tpl.html and unit-alignment-edit-modal.tpl.html) to use the new Angular syntax, and I registered the directive in doubtfire-angular.module.ts. Looking through your code. It is well structured and commented. I had no issues running it on my end, although I am sure I reviewed something similar to this already. Thank you for the opportunity to review your work

ibi420 commented:
Hello @aditya993388, In this task, you have migrated the old contenteditable directive from AngularJS to Angular by creating a new directive file (content-editable.directive.ts) using ControlValueAccessor for [(ngModel)] support. You updated two template files (unit-ilo-edit-modal.tpl.html and unit-alignment-edit-modal.tpl.html) to use the new Angular syntax and registered the directive in doubtfire-angular.module.ts. Looking through your code. It is well structured and commented. I had no issues running it on my end, although I am sure I reviewed something similar to this already. Thank you for the opportunity to review your work

aditya993388 committed:
Changes after review

martindolores reviewed (approved):
Great work on this migration! You were able to migrate the content-editable directive to Angular and still maintain support for AngularJS components. Nice work! 

Keep in mind there is another existing PR that does the same migration here: https://github.com/thoth-tech/doubtfire-web/pull/310

ibi420 commented:
Hi @aditya993388 thanks for clarifying the difference in the migration of a component and a directive. I missed the difference earlier and checked the usual migration steps for a component migration. After learning more about it, I have reviewed your migration again and I had no issues running it on my end. Looking through your code. It is well structured and commented. Thank you for the opportunity to review your work.
